File: python-behave/pages/project_page.py
"""
Project page object for the Kanban board.

Handles navigation between projects and task verification.
"""
from typing import List
import re
from playwright.sync_api import Page, Locator
from .base_page import BasePage
from core.logger import get_logger

logger = get_logger(__name__)


class ProjectPage(BasePage):
    """Page object for the Kanban project board."""

    # Selectors
    SIDEBAR = 'aside, nav, [class*="sidebar"]'
    PROJECT_BUTTON = 'button'
    COLUMN_HEADER = 'h2'
    TASK_CARD = '[class*="bg-white"]'

    # Tag detection
    TAG_MAX_LENGTH = 20
    DATE_PATTERN = re.compile(r'\d{1,2}/\d{1,2}/\d{4}')

    @property
    def url_path(self) -> str:
        return "/"

    # --- Navigation ---

    def navigate_to_project(self, project_name: str) -> "ProjectPage":
        """
        Navigate to a specific project by clicking in sidebar.

        Args:
            project_name: Name of the project to navigate to

        Returns:
            Self for method chaining
        """
        logger.info(f"Navigating to project: {project_name}")

        # Find and click the project button in sidebar
        sidebar = self.page.locator(self.SIDEBAR).first
        project_button = sidebar.locator(self.PROJECT_BUTTON).filter(has_text=project_name).first
        project_button.click()

        # Wait for board to load
        self.page.wait_for_load_state("networkidle")

        return self

    def get_visible_projects(self) -> List[str]:
        """Get list of visible project names in sidebar."""
        sidebar = self.page.locator(self.SIDEBAR).first
        buttons = sidebar.locator(self.PROJECT_BUTTON).all()
        projects = []
        for button in buttons:
            # Extract the primary heading from the button
            heading = button.locator("h2, span, div").first
            if heading.count() > 0:
                text = heading.text_content()
            else:
                text = button.text_content()
            if text and text.strip():
                projects.append(text.strip())
        return projects

    def get_current_project_name(self) -> str:
        """Get the currently selected project name from header."""
        # Target the project header specifically (text-xl class), not the sidebar "Projects" h1
        header = self.page.locator('h1[class*="text-xl"]').first
        if header.is_visible():
            return header.text_content() or ""
        return ""

    # --- Column Operations ---

    def get_column(self, column_name: str) -> Locator:
        """
        Get a column locator by name.

        Args:
            column_name: Column name (To Do, In Progress, Review, Done)

        Returns:
            Locator for the column container
        """
        # Find column header, then locate the nearest container within main content
        column_header = self.page.locator(self.COLUMN_HEADER).filter(has_text=column_name).first
        main_content = self.page.locator("main, [class*='content'], [class*='board']").first

        return main_content.locator("div").filter(has=column_header).first

    def get_column_names(self) -> List[str]:
        """Get list of all column names."""
        headers = self.page.locator(self.COLUMN_HEADER).all()
        column_names = []
        for header in headers:
            text = header.text_content() or ""
            normalized = re.sub(r"\s*\(\d+\)\s*$", "", text).strip()
            # Filter out non-column headers (might match task titles)
            if normalized in ["To Do", "In Progress", "Review", "Done"]:
                column_names.append(normalized)
        return column_names

    def get_column_task_count(self, column_name: str) -> int:
        """
        Get task count from column header display.

        Column headers show format: "Column Name (N)" e.g., "To Do (2)".
        Reading the count from the header is more reliable than counting
        card elements which may match non-card containers.

        Args:
            column_name: Column name

        Returns:
            Number of tasks shown in the column header
        """
        column_header = self.page.locator(self.COLUMN_HEADER).filter(
            has_text=column_name
        ).first
        text = column_header.text_content() or ""
        match = re.search(r'\((\d+)\)', text)
        return int(match.group(1)) if match else -1

    # --- Task Operations ---

    def _get_task_card(self, task_name: str, column_name: str) -> Locator:
        """
        Get a task card locator within a specific column.

        Uses column scoping with exact task title text to find the card container.
        Note: The has= filter requires a page-scoped (not column-scoped) locator
        to work correctly as a relative match within each candidate element.
        """
        column = self.get_column(column_name)
        task_title = self.page.get_by_text(task_name, exact=True)

        return column.locator(self.TASK_CARD).filter(has=task_title).first

    def verify_task_in_column(self, task_name: str, column_name: str) -> "ProjectPage":
        """
        Verify a task exists in the specified column.

        Args:
            task_name: Name of the task
            column_name: Expected column name

        Returns:
            Self for method chaining

        Raises:
            AssertionError: If task not found in column
        """
        logger.info(f"Verifying task '{task_name}' in column '{column_name}'")

        column = self.get_column(column_name)
        task = column.get_by_text(task_name, exact=True)

        # Assert task is visible in the column
        task.wait_for(state="visible", timeout=self.config.default_timeout)

        return self

    def get_task_tags(self, task_name: str, column_name: str) -> List[str]:
        """
        Get tags for a specific task.

        Args:
            task_name: Name of the task
            column_name: Column containing the task

        Returns:
            List of tag names
        """
        card = self._get_task_card(task_name, column_name)

        # Tags are colored spans using Tailwind bg-* classes
        tag_selectors = [
            'span[class*="bg-"]',       # Tailwind background color (primary)
            'span[class*="tag"]',
            'span[class*="badge"]',
            'span[class*="rounded"]',
        ]

        tags = []
        for selector in tag_selectors:
            tag_elements = card.locator(selector).all()
            for tag in tag_elements:
                text = tag.text_content()
                if text and text.strip():
                    trimmed = text.strip()
                    # Filter: tags are short text, exclude dates
                    if (len(trimmed) < self.TAG_MAX_LENGTH
                            and not self.DATE_PATTERN.search(trimmed)
                            and '/' not in trimmed):
                        tags.append(trimmed)

        # Deduplicate while preserving order
        seen = set()
        unique_tags = []
        for tag in tags:
            if tag not in seen:
                seen.add(tag)
                unique_tags.append(tag)

        return unique_tags

    def verify_task_tags(
        self, task_name: str, expected_tags: List[str], column_name: str
    ) -> "ProjectPage":
        """
        Verify a task has the expected tags.

        Args:
            task_name: Name of the task
            expected_tags: List of expected tag names
            column_name: Column containing the task

        Returns:
            Self for method chaining

        Raises:
            AssertionError: If tags don't match
        """
        logger.info(f"Verifying tags for task '{task_name}': {expected_tags}")

        actual_tags = self.get_task_tags(task_name, column_name)

        for expected_tag in expected_tags:
            if expected_tag not in actual_tags:
                raise AssertionError(
                    f"Tag '{expected_tag}' not found for task '{task_name}'. "
                    f"Found tags: {actual_tags}"
                )

        return self

    # --- Sidebar Operations ---

    def is_sidebar_visible(self) -> bool:
        """Check if sidebar is visible."""
        return self.is_visible(self.SIDEBAR, timeout=5000)

File: src/pages/project.page.ts
import { Locator, expect } from '@playwright/test';
import { BasePage } from './base.page.js';
import { TAG_MAX_LENGTH } from '../utils/config.js';

/**
 * Regular expression pattern for matching column headers.
 * Columns display as "Column Name (count)" e.g., "To Do (2)"
 */
const COLUMN_HEADER_PATTERN = (columnName: string): RegExp =>
  new RegExp(`^${columnName}\\s*\\(\\d+\\)$`);

/**
 * Date pattern to exclude from tag detection.
 * Matches dates like "3/24/2024" or "12/31/2024"
 */
const DATE_PATTERN = /\d{1,2}\/\d{1,2}\/\d{4}/;

/**
 * Page object for the project board (Kanban) page.
 * Handles navigation between projects and task/tag verification.
 *
 * App Structure:
 * - Left sidebar: Dark background with project buttons
 * - Main content: Kanban columns (To Do, In Progress, Review, Done)
 * - Task cards: Title, description, colored tags, assignee, date
 *
 * Selector Strategy:
 * - Project buttons: Filter sidebar buttons by text content
 * - Columns: Match header text pattern, traverse to container
 * - Tasks: Filter cards by text content within column scope
 * - Tags: Colored spans within task cards, excluding dates
 */
export class ProjectPage extends BasePage {
  public readonly path = '/';

  /**
   * Get the sidebar element containing project navigation.
   */
  private get sidebar(): Locator {
    return this.page.locator('aside, [class*="sidebar"], nav').first();
  }

  /**
   * Get a project button from the sidebar.
   *
   * @param projectName - The name of the project to find
   * @returns Locator for the project button
   */
  private getProjectButton(projectName: string): Locator {
    return this.sidebar.locator('button').filter({ hasText: projectName }).first();
  }

  /**
   * Get the header element for a column.
   *
   * @param columnName - The name of the column (e.g., "To Do")
   * @returns Locator for the column header text
   */
  private getColumnHeader(columnName: string): Locator {
    return this.page.getByText(COLUMN_HEADER_PATTERN(columnName));
  }

  /**
   * Get the main content area (excludes sidebar).
   * This is where the Kanban columns are displayed.
   */
  private get mainContent(): Locator {
    return this.page.locator('main, [class*="content"], [class*="board"]').first();
  }

  /**
   * Get the container element for a column.
   * Finds a column container that has the matching header text.
   *
   * Approach: Look for div containers in the main content area
   * that contain the column header text. This is more robust than
   * XPath ancestor traversal as it uses Playwright's built-in filtering.
   *
   * @param columnName - The name of the column
   * @returns Locator for the column container
   */
  private getColumn(columnName: string): Locator {
    // Find a container div that has the column header
    // The column header pattern includes count like "To Do (2)"
    return this.mainContent
      .locator('div')
      .filter({ has: this.page.getByText(COLUMN_HEADER_PATTERN(columnName)) })
      .first();
  }

  /**
   * Get a task card within a specific column.
   * Uses the column container to scope the search.
   *
   * @param taskName - The name/title of the task
   * @param columnName - The column to search within
   * @returns Locator for the task card within the column
   */
  private getTaskInColumn(taskName: string, columnName: string): Locator {
    const column = this.getColumn(columnName);

    // Find the task title within the column, then locate its parent card
    // This ensures we get the exact card containing this specific task
    const taskTitle = column.getByText(taskName, { exact: true });

    // Navigate up to the card container (parent divs with card-like classes)
    return taskTitle.locator('xpath=ancestor::div[contains(@class, "bg-white") or contains(@class, "rounded")][1]');
  }

  /**
   * Get tag elements within a task card.
   * Tags are colored spans, excluding date elements.
   *
   * @param taskCard - The task card locator
   * @returns Locator for tag elements
   */
  private getTaskTags(taskCard: Locator): Locator {
    return taskCard
      .locator(
        'span[class*="bg-"], span[class*="badge"], span[class*="tag"], span[class*="rounded"]'
      )
      .filter({ hasNotText: DATE_PATTERN });
  }

  /**
   * Get all project names visible in the sidebar.
   *
   * @returns Array of project name strings
   */
  public async getVisibleProjects(): Promise<string[]> {
    const buttons = this.sidebar.locator('button');
    const count = await buttons.count();
    const projects: string[] = [];
    for (let i = 0; i < count; i++) {
      // Project buttons contain the name and a subtitle — extract the h2 text
      const nameEl = buttons.nth(i).locator('h2, span, div').first();
      const hasName = await nameEl.count();
      let text: string;
      if (hasName > 0) {
        text = (await nameEl.textContent()) ?? '';
      } else {
        text = (await buttons.nth(i).textContent()) ?? '';
      }
      if (text.trim()) {
        projects.push(text.trim());
      }
    }
    return projects;
  }

  /**
   * Get all visible column names on the current board.
   *
   * @returns Array of column name strings (without counts)
   */
  public async getColumnNames(): Promise<string[]> {
    const columns: string[] = [];
    for (const name of ['To Do', 'In Progress', 'Review', 'Done']) {
      const header = this.getColumnHeader(name);
      if ((await header.count()) > 0 && (await header.isVisible())) {
        columns.push(name);
      }
    }
    return columns;
  }

  /**
   * Get the task count displayed in a column header.
   * Column headers show format: "Column Name (N)"
   *
   * @param columnName - The column to check
   * @returns The numeric count from the header
   */
  public async getColumnTaskCount(columnName: string): Promise<number> {
    const header = this.getColumnHeader(columnName);
    await expect(header, `Column "${columnName}" header should be visible`).toBeVisible();
    const text = await header.textContent();
    const match = text?.match(/\((\d+)\)/);
    return match ? parseInt(match[1], 10) : -1;
  }

  /**
   * Get the current project name from the main content header.
   * The project header is an h1 with text-xl styling, distinct from sidebar headers.
   *
   * @returns The project name displayed in the header
   */
  public async getCurrentProjectName(): Promise<string> {
    const header = this.page.locator('h1[class*="text-xl"]').first();
    const text = await header.textContent();
    return text?.trim() ?? '';
  }

  /**
   * Verify page is loaded and user is authenticated.
   * Used by the fixture as a sanity check.
   *
   * @throws Error if user appears to not be authenticated
   */
  public async verifyPageLoaded(): Promise<void> {
    await this.page.waitForLoadState('domcontentloaded');

    // Verify we see the Projects header in sidebar
    const projectsHeader = this.page.getByText('Projects');
    await expect(projectsHeader, 'Projects header should be visible').toBeVisible({
      timeout: 10000,
    });
  }

  /**
   * Navigate to a specific project by clicking its sidebar button.
   *
   * @param projectName - The name of the project to navigate to
   */
  public async navigateToProject(projectName: string): Promise<void> {
    const projectButton = this.getProjectButton(projectName);

    await expect(projectButton, `Project button "${projectName}" should be visible`).toBeVisible();
    await projectButton.click();

    await this.page.waitForLoadState('domcontentloaded');

    // Verify the project header shows the selected project
    const projectHeader = this.page.locator('h1, h2').filter({ hasText: projectName }).first();
    await expect(
      projectHeader,
      `Project "${projectName}" should be displayed as header`
    ).toBeVisible();
  }

  /**
   * Verify a task exists in the expected column.
   *
   * @param taskName - The name of the task to find
   * @param columnName - The expected column name
   */
  public async verifyTaskInColumn(taskName: string, columnName: string): Promise<void> {
    // First verify the column exists
    const columnHeader = this.getColumnHeader(columnName);
    await expect(columnHeader, `Column "${columnName}" should exist`).toBeVisible();

    // Verify task is in the correct column
    const taskInColumn = this.getTaskInColumn(taskName, columnName);
    await expect(
      taskInColumn,
      `Task "${taskName}" should be in column "${columnName}"`
    ).toBeVisible({ timeout: 10000 });
  }

  /**
   * Get all visible tag names from a task card.
   *
   * @param taskName - The name of the task
   * @param columnName - The column containing the task
   * @returns Array of tag text values
   */
  public async getTaskTagNames(taskName: string, columnName: string): Promise<string[]> {
    const taskCard = this.getTaskInColumn(taskName, columnName);
    await expect(taskCard, `Task card "${taskName}" should be visible`).toBeVisible();

    const tags = this.getTaskTags(taskCard);
    const tagCount = await tags.count();

    const tagNames: string[] = [];
    for (let i = 0; i < tagCount; i++) {
      const tagText = await tags.nth(i).textContent();
      if (tagText?.trim()) {
        const trimmed = tagText.trim();
        // Filter: tags are short text without date separators
        if (trimmed.length < TAG_MAX_LENGTH && !trimmed.includes('/')) {
          tagNames.push(trimmed);
        }
      }
    }

    return tagNames;
  }

  /**
   * Verify a task has all expected tags.
   * Tag comparison is case-insensitive.
   *
   * @param taskName - The name of the task
   * @param expectedTags - Array of expected tag names
   * @param columnName - The column containing the task
   */
  public async verifyTaskTags(
    taskName: string,
    expectedTags: string[],
    columnName: string
  ): Promise<void> {
    const actualTags = await this.getTaskTagNames(taskName, columnName);

    const normalize = (value: string): string => value.trim().toLowerCase();
    const actualSet = new Set(actualTags.map(normalize));
    const expectedSet = new Set(expectedTags.map(normalize));

    for (const expectedTag of expectedSet) {
      const hasTag = actualSet.has(expectedTag);

      expect(
        hasTag,
        `Task "${taskName}" should have tag "${expectedTag}". Actual tags found: [${actualTags.join(', ')}]`
      ).toBe(true);
    }

    expect(
      actualSet.size,
      `Task "${taskName}" should have only expected tags. Actual tags: [${actualTags.join(', ')}], Expected tags: [${expectedTags.join(', ')}]`
    ).toBe(expectedSet.size);
  }
}
